doc/developer/design/20240411_graceful_reconfiguration_of_managed_clusters_v1.md

```diff
@@ -0,0 +1,367 @@
+# Graceful Reconfiguration of Managed Clusters
+
+Associated issues/prs:
+[#20010](https://github.com/MaterializeInc/materialize/issues/20010)
+[#26401](https://github.com/MaterializeInc/materialize/pull/26401)
+
+## The Problem
+Reconfiguring a manage dcluster via `alter cluster...` leads to downtime. This
+is due to our cluster sequencer first removing the old replicas then adding new
+replicas that match the new configuration. The duration of downtime can be seen
+as extending through the full period it takes to rehydrate the new replicas.
+
+## Success Criteria
+Provide a mechanism which allows user to delay deletion of old replicas for
+some user defined duraiton while the new replicas spin up. This mechanism should
+allow for a follow on feature, where teardown is delayed until the newly created
+replicas have finished hydrating, or reach some parity with the replicas they
+are replacing.
+
+## V1 Scope
+ - Timeout based reconfiguration cleanup scheduling
```

Comment:
I guess I see the timeout as bounds you're setting on the cost of zero downtime during reconfiguration and hydration-detection as our way of helping reduce those costs.

Reply:
> Is this useful for users? It will still lead to unpredictable downtime. (Perhaps some of this has been discussed in places I'm not aware of but:) I'm interested in not building the timeout feature and only building the hydration-detection feature.

The `WAIT = FOR <interval>` is definitely not the ideal user experience, but I think we can ship it much faster than we can ship `WAIT = UNTIL CAUGHT UP`. No one has yet proposed a clear implementation for computing "is this replica caught up?", and I'd hate to block this work 

Maybe the right goal is to aim for getting `WAIT = FOR <interval>` merged and in "private preview", with the expectation that it's unlikely that folks will want to use it. But we can at least check our assumption there! And while we're getting feedback on `WAIT = FOR <interval>` we can try to shake out the question of computing "is this replica caught up?", and it should drop into place pretty cleanly with the foundation we lay with `WAIT = FOR< interval>`.